x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -0,0 +1,280 @@
+                for(String modelKey : loadingListeners.keySet()) {
+                    if (allReferencedModelKeys.contains(modelKey) == false) {
+                        drainWithFailure.addAll(loadingListeners.remove(modelKey));
+                    }
+                }
+
+                // Remove all cached models that are not referenced by any processors
+                loadedModels.keySet().retainAll(allReferencedModelKeys);
+
+                // Remove all that are currently being loaded
+                allReferencedModelKeys.removeAll(loadingListeners.keySet());
+
+                // Remove all that are fully loaded, will attempt empty model loading again
+                loadedModels.forEach((id, optionalModel) -> {
+                    if(optionalModel.isSuccess()) {
+                        allReferencedModelKeys.remove(id);
+                    }
+                });
+                // Populate loadingListeners key so we know that we are currently loading the model
+                for(String modelId : allReferencedModelKeys) {
+                    loadingListeners.put(modelId, new ArrayDeque<>());
+                }
+            }
+            for(ActionListener<Model> listener = drainWithFailure.poll(); listener != null; listener = drainWithFailure.poll()) {
+                listener.onFailure(
+                    new ElasticsearchException("Cancelling model load and inference as it is no longer referenced by a pipeline"));
+            }
+            loadModels(allReferencedModelKeys);
+        }
+    }
+
+    private void loadModels(Set<String> modelKeys) {
+        if (modelKeys.isEmpty()) {
+            return;
+        }
+        // Execute this on a utility thread as when the callbacks occur we don't want them tying up the cluster listener thread pool
+        threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(() -> {
+            for(String modelKey : modelKeys) {
+                Tuple<String, Long> modelIdAndVersion = splitModelKey(modelKey);
+                this.loadModel(modelKey, modelIdAndVersion.v1(), modelIdAndVersion.v2());
+            }
+        });
+    }
+
+    private static <T> Queue<T> addFluently(Queue<T> queue, T object) {
+        queue.add(object);
+        return queue;
+    }
+
+    private static String modelKey(String modelId, long modelVersion) {
+        return modelId + "_" + modelVersion;
+    }
+
+    private static Tuple<String, Long> splitModelKey(String modelKey) {
+        int delim = modelKey.lastIndexOf('_');
+        String modelId = modelKey.substring(0, delim);
+        Long version = Long.valueOf(modelKey.substring(delim + 1));
+        return Tuple.tuple(modelId, version);
+    }
+
+    private static Set<String> getReferencedModelKeys(IngestMetadata ingestMetadata) {
+        Set<String> allReferencedModelKeys = new HashSet<>();
+        if (ingestMetadata != null) {
+            ingestMetadata.getPipelines().forEach((pipelineId, pipelineConfiguration) -> {
```

Comment:
@dimitris-athanasiou sadly, I don't think this is possible. 

The `IngestService` requires a view of all the loaded Ingest plugins (of which ML will be one). 

Since `ModelLoadingService` is built within our `createComponents` method, it does not have access to a constructed IngestService anywhere. 

I looked into how to create the `processorFactories` parameter myself, but again, it requires a view of all the loaded `IngestPlugin` classes. Again, I cannot find a place where we have access to the complete list of loaded plugins on the node in the `createComponents` method.

I think to support this type of thing, we will have to either:

A. Inject the IngestService into the `createComponents` method
B. Inject the list of loaded plugins into the `createComponents` method

It seems to me that A is the least invasive, but would require an update to a core method that every plugin uses...not sure it is worth it.

Reply:
Sad indeed. Thanks for looking into it!